src/ore/src/id_gen.rs

```diff
@@ -41,66 +47,114 @@ impl<Id: From<u64> + Default> Gen<Id> {
+    + serde::Serialize
+    + fmt::Display
+{
+}
+
+impl<T> IdGenerator for T where
+    T: From<u8>
+        + AddAssign
+        + Sub
+        + PartialOrd
+        + Copy
+        + Eq
+        + Hash
+        + Ord
+        + rand::distributions::uniform::SampleUniform
+        + serde::Serialize
+        + fmt::Display
+{
+}
+
 /// Manages allocation of numeric IDs.
-///
-/// Note that the current implementation wastes memory. It would be far more
-/// efficient to use a compressed bitmap, like <https://roaringbitmap.org> or
-/// the hibitset crate, but neither presently supports a fast "find first zero"
-/// operation.
 #[derive(Debug, Clone)]
-pub struct IdAllocator<T>(Arc<Mutex<IdAllocatorInner<T>>>);
+pub struct IdAllocator(Arc<Mutex<IdAllocatorInner>>);
 
 #[derive(Debug)]
-struct IdAllocatorInner<T> {
-    next: T,
-    max: T,
-    free: VecDeque<T>,
+struct IdAllocatorInner {
+    next: StdRng,
+    min: u32,
+    max: u32,
+    used: BitSet,
 }
 
-impl<T> IdAllocator<T>
-where
-    T: From<u8> + AddAssign + PartialOrd + Copy,
-{
+impl IdAllocator {
     /// Creates a new `IdAllocator` that will assign IDs between `min` and
     /// `max`, both inclusive.
-    pub fn new(min: T, max: T) -> IdAllocator<T> {
+    pub fn new(min: u32, max: u32) -> IdAllocator {
+        assert!(min <= max);
+        let total = usize::cast_from(max - min);
+        assert!(total < BitSet::BITS_PER_USIZE.pow(4));
         let inner = IdAllocatorInner {
-            next: min,
+            next: StdRng::from_entropy(),
+            min,
             max,
-            free: VecDeque::new(),
+            used: BitSet::new(),
         };
         IdAllocator(Arc::new(Mutex::new(inner)))
     }
 
-    /// Allocates a new ID.
+    /// Allocates a new ID randomly distributed between min and max.
     ///
     /// Returns `None` if the allocator is exhausted.
     ///
     /// The ID associated with the [`IdHandle`] will be freed when all of the
     /// outstanding [`IdHandle`]s have been dropped.
-    pub fn alloc(&self) -> Option<IdHandle<T>> {
+    pub fn alloc(&self) -> Option<IdHandle<u32>> {
         let inner = Arc::new(internal::IdHandleInner::new(self)?);
         Some(IdHandle::Dynamic(inner))
     }
 
-    fn alloc_internal(&self) -> Option<T> {
+    // Attempt to allocate a new ID. We want the ID to be randomly distributed in the range. To do
+    // this, choose a random candidate. Check if it's already in the used set, and increment in a
+    // loop until an unused slot is found. Ideally we could ask used set what its next used or
+    // unused id is after some given X, but that's not part of the API. This means that when the
+    // range is large and the used set is near full, we will spend a lot of cycles looking for an
+    // open slot. However, we limit the number of connections to the thousands, and connection IDs
+    // have 20 bits (~1 million) of space, so it is not currently possible to enter that state.
```

Comment:
Just a thought, what if we tracked how many IDs we've allocated thus far, and generated a number `x` in the range of `[min, max - len]`. Then get the `x`th number in the inverse of `used`?

I _think_ a CPU efficient way to do this using a [roaring bitmap](https://docs.rs/roaring/latest/roaring/bitmap/struct.RoaringBitmap.html) that we fill with values `[min, max]`, this would represent the list of available ids. When allocating one we `remove(...)` it from the set, otherwise we can `select(x)`. I _think_ `select(x)` runs in logarithmic time? I could be off base here, I don't know how roaring bitmaps work super well.

Reply:
I had excluded roaring bitmaps from consideration due to their docs about sparse data:

> The sparse scenario is another use case where compressed bitmaps should not be used. Keep in mind that random-looking data is usually not compressible. E.g., if you have a small set of 32-bit random integers, it is not mathematically possible to use far less than 32 bits per integer, and attempts at compression can be counterproductive.

I think we might need to collect some data here, and compare the memory and cpu of roaring, hibitset, and HashSet<u32>.

Reply:
I've only skimmed, but `hi_sparse_bitset` looks like it gives us what we want: https://crates.io/crates/hi_sparse_bitset. In particular, see the ability to create a cursor from an index: https://docs.rs/hi_sparse_bitset/0.4.0/hi_sparse_bitset/iter/struct.IndexCursor.html#method.from.

It has a nice comparison of the options, too: https://crates.io/crates/hi_sparse_bitset